src/quick-lint-js/fe/parse-expression.cpp

```diff
@@ -85,6 +85,12 @@ void parser::visit_expression(expression* ast, parse_visitor_base& v,
     expression* lhs = ast->child_0();
     expression* rhs = ast->child_1();
     this->visit_assignment_expression(lhs, rhs, v);
+
+    if (lhs->span().string_view() == rhs->span().string_view()) {
```

Comment:
Must fix: Here, you are checking that the source code of the left and right side have the same text. This means that `x.p = x.p;` will report `diag_variable_assigned_to_self_is_noop`. It also means that `x = \u{78};` (which is equivalent to `x = x;`) will not report a diagnostic, even though it should.

To fix this, I would check that both `lhs` and `rhs` have `expression_kind::variable`. If so, check `lhs->variable_identifier().normalized_name() == rhs->variable_identifier().normalized_name()`.

It might be a good idea to also handle `(x) = x;` and `x = ((x));`. `expression::without_paren` might be handy.